### PR TITLE
Core/SAI: Implement quest source for SAI

### PR DESCRIPTION
**Pull request by matanshukry**
**Changes proposed:**

-  Implement quest source for SAI

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:**

**Tests performed:**

(Does it build, tested in-game, etc.)

- [x] It builds
- [x] Tested in-game

**Known issues and TODO list:**

**Use case:**

Pushed quests (such as wqs) where conversations happens when a quest has been given or completed to player

**Example Usage**

Currently if you level a character to 48 you'll get the quest,
so this quest will teleport you to stormwind once the player has that quest.

```sql
SET @QUEST_SL := 60545; -- A chilling summon

DELETE FROM `smart_scripts` WHERE `entryorguid` = @QUEST_SL AND `source_type` = 5;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`event_param5`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(@QUEST_SL, 5, 0, 1, 47, 0, 100, 0, 0, 0, 0, 0, 0, 62, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, -8833.07, 622.778, 93.9317, 0.6771, "On QUEST_SL accepted - teleport to stormwind");

DELETE FROM `quest_template_addon` WHERE `ID` = @QUEST_SL;
INSERT INTO `quest_template_addon` (`ID`, `ScriptName`) VALUES
(@QUEST_SL, 'SmartQuest');
```

### CODE REVIEW

**Comment by offl** [comment]
Great
Did you tried to use action lists with different target types(quest-giver 1 and player-invoker 7)?

@jackpoz maybe you want to port this to 3.3.5?

**Comment by matanshukry** [comment]
> Did you tried to use action lists with different target types(quest-giver 1 and player-invoker 7)?

I didn't, but I believe it would only work if the events are linked.
I think those restrictions also applies to scene and area triggers though?

Also, currently (before this PR) you can still handle events on quests using a specific NPC and quest id, this PR is needed for pushed quests where there is no quest giver. Is there anyway to have multiple targets in that case?

**Comment by offl** [comment]
Yes, you're right, areatriggers can't load action list but can use linked events. I thought initially (https://github.com/TrinityCore/TrinityCore/issues/24961#issuecomment-726807474) SMART_SCRIPT_TYPE_QUEST, SMART_SCRIPT_TYPE_GOSSIP and SMART_SCRIPT_TYPE_EVENT are supposed to act as event holders, like creatures or objects do. So they can load action lists and target type 1 (self) will refer to event holder(quest-giver, gossip-menu owner)
But looks like it requires more changes

**Comment by jackpoz** [comment]
> @jackpoz maybe you want to port this to 3.3.5?

@offl sure but only after it has been merged to master

**Comment by offl** [comment]
>Yes, you're right, areatriggers can't load action list but can use linked events. I thought initially (#24961 (comment)) SMART_SCRIPT_TYPE_QUEST, SMART_SCRIPT_TYPE_GOSSIP and SMART_SCRIPT_TYPE_EVENT are supposed to act as event holders, like creatures or objects do. So they can load action lists and target type 1 (self) will refer to event holder(quest-giver, gossip-menu owner)
But looks like it requires more changes

Technically we can load an action list via SMART_SCRIPT_TYPE_QUEST or SMART_SCRIPT_TYPE_GOSSIP to quest-giver and gossip-owner. In some cases it's exactly what those script types are supposed to do. Benefits: we are avoinding to script quest taken\rewarded events or gossip select events in main script because it's ugly and may cause problems in future.
For example we want to convert all scripts to free ids predefined templates like Blizz did it. Let's say we want to convert all combat actions to predefined templates, but what's left? Exactly, all those gossip scripts which are supposed to be linked to gossip entry, all quest scripts originally linked to quest entries. All waypoints scripts originally linked to point ids(biggest problem ever). Spellhit events are kinda combat actions but mostly all of them can be linked to script ids(btw SAI also can legitimately replace spell_linked_spell which is indeed hack but not because those scripts are scripted in DB but because that system is not how it's supposed to work. There are 700 linked spells in 3.3.5 and I really doubt someone wants to make 700 scripts just to cast another spell after OnRemove or AfterCast. Spell scripts like creature scripts should be used only to make something complicated, like boss scripts or player spells. To cast another spell or cast random spell you don't need to create C++ spell script because it's insane).

However we have also SMART_SCRIPT_TYPE_EVENT and that thing is more complex because they are map scripts and can be updated even if event starter is dead or despawned(lots of events are called when object despawns). At least it's how event_scripts always worked. In some cases there is no creature or gameobject invoker(taxi events)

So SMART_SCRIPT_TYPE_QUEST or SMART_SCRIPT_TYPE_GOSSIP too should be able to call action lists to quest-giver or gossip-owner otherwise those scripts cannot be used the way they are used in retail
Just thoughts, I'm not asking to implement that thing

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -3815,6 +3832,12 @@ void SmartScript::OnInitialize(WorldObject* obj, AreaTriggerEntry const* at, Sce
         sceneTemplate = scene;
         TC_LOG_DEBUG("scripts.ai", "SmartScript::OnInitialize: source is Scene with id %u", scene->SceneId);
     }
+    else if (qst)
+    {
+        mScriptType = SMART_SCRIPT_TYPE_QUEST;
+        quest = qst;
+        TC_LOG_DEBUG("scripts.ai", "SmartScript::OnInitialize: source is Quest with id %u", scene->SceneId);
```
Comment (Shauren) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:3839]: Will crash if debug logger is enabled (using nullptr scene pointer instead of quest)

**Review by Shauren — CHANGES_REQUESTED** [review]

**Pull request merged**